sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala

```diff
@@ -0,0 +1,328 @@
+        CUSTOM_METRIC_FILES_COPIED -> rocksDBMetrics.filesCopied,
+        CUSTOM_METRIC_FILES_REUSED -> rocksDBMetrics.filesReused
+      ) ++ rocksDBMetrics.zipFileBytesUncompressed.map(bytes =>
+        Map(CUSTOM_METRIC_ZIP_FILE_BYTES_UNCOMPRESSED -> bytes)).getOrElse(Map())
+
+      StateStoreMetrics(
+        rocksDBMetrics.numUncommittedKeys,
+        rocksDBMetrics.memUsageBytes,
+        stateStoreCustomMetrics)
+    }
+
+    override def hasCommitted: Boolean = state == COMMITTED
+
+    override def toString: String = {
+      s"RocksDBStateStore[id=(op=${id.operatorId},part=${id.partitionId})," +
+        s"dir=${id.storeCheckpointLocation()}]"
+    }
+
+    /** Return the [[RocksDB]] instance in this store. This is exposed mainly for testing. */
+    def dbInstance(): RocksDB = rocksDB
+  }
+
+  override def init(
+      stateStoreId: StateStoreId,
+      keySchema: StructType,
+      valueSchema: StructType,
+      indexOrdinal: Option[Int],
+      storeConf: StateStoreConf,
+      hadoopConf: Configuration): Unit = {
+    this.stateStoreId_ = stateStoreId
+    this.keySchema = keySchema
+    this.valueSchema = valueSchema
+    this.storeConf = storeConf
+    this.hadoopConf = hadoopConf
+    rocksDB // lazy initialization
+  }
+
+  override def stateStoreId: StateStoreId = stateStoreId_
+
+  override def getStore(version: Long): StateStore = {
+    require(version >= 0, "Version cannot be less than 0")
+    rocksDB.load(version)
+    new RocksDBStateStore(version)
+  }
+
+  override def doMaintenance(): Unit = {
+    rocksDB.cleanup()
+  }
+
+  override def close(): Unit = {
+    rocksDB.close()
+  }
+
+  override def supportedCustomMetrics: Seq[StateStoreCustomMetric] = ALL_CUSTOM_METRICS
+
+  private[state] def latestVersion: Long = rocksDB.getLatestVersion()
+
+  /** Internal fields and methods */
+
+  @volatile private var stateStoreId_ : StateStoreId = _
+  @volatile private var keySchema: StructType = _
+  @volatile private var valueSchema: StructType = _
+  @volatile private var storeConf: StateStoreConf = _
+  @volatile private var hadoopConf: Configuration = _
+
+  private[sql] lazy val rocksDB = {
+    val dfsRootDir = stateStoreId.storeCheckpointLocation().toString
+    val storeIdStr = s"StateStoreId(opId=${stateStoreId.operatorId}," +
+      s"partId=${stateStoreId.partitionId},name=${stateStoreId.storeName})"
+    val sparkConf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf)
+    val localRootDir = Utils.createTempDir(Utils.getLocalDir(sparkConf), storeIdStr)
+    new RocksDB(dfsRootDir, RocksDBConf(storeConf), localRootDir, hadoopConf, storeIdStr)
+  }
+
+  private lazy val encoder = new StateEncoder
+
+  private def verify(condition: => Boolean, msg: String): Unit = {
+    if (!condition) { throw new IllegalStateException(msg) }
+  }
+
+  /**
+   * Encodes/decodes UnsafeRows to versioned byte arrays.
+   * It uses the first byte of the generated byte array to store the version the describes how the
```

Comment:
I guess this is a typo. "the" -> "that".

Reply:
Ah yea, thanks!